### ISSUE TITLE
When call vkDestroyInstance the launch app will crash

### ISSUE THREAD

**Issue by FuXiii** [baldurk/renderdoc#2859]
## Description

When I use `RenderDoc V1.25` to run `Vulkan` app, the app will crash, but if I independent execution the `Vulkan` app it will execute normally. then I try to find where is source of the error. 
## Steps to reproduce

The `Vulkan` app will run the code:
```CXX
bool Turbo::Core::TInstance::IsSupportVulkan()
{
    std::cout << "0.0.0" << std::endl;

    TVulkanLoader::Instance();//it will call LoadLibraryA("vulkan-1.dll") and GetProcAddress(library, "vkGetInstanceProcAddr") to get vkGetInstanceProcAddr function

    std::cout << "0.0.1" << std::endl;

    VkApplicationInfo application_info = {};
    application_info.sType = VkStructureType::VK_STRUCTURE_TYPE_APPLICATION_INFO;
    application_info.pNext = nullptr;
    application_info.pApplicationName = nullptr;
    application_info.applicationVersion = 0;
    application_info.pEngineName = nullptr;
    application_info.engineVersion = 0;
    application_info.apiVersion = VK_MAKE_VERSION(1, 0, 0);

    VkInstanceCreateInfo vkinstance_create_info = {};
    vkinstance_create_info.sType = VkStructureType::VK_STRUCTURE_TYPE_INSTANCE_CREATE_INFO;
    vkinstance_create_info.pNext = nullptr;
    vkinstance_create_info.flags = 0;
    vkinstance_create_info.pApplicationInfo = &application_info;
    vkinstance_create_info.enabledLayerCount = 0;
    vkinstance_create_info.ppEnabledLayerNames = nullptr;
    vkinstance_create_info.enabledExtensionCount = 0;
    vkinstance_create_info.ppEnabledExtensionNames = nullptr;

    VkInstance instance = VK_NULL_HANDLE;

    VkAllocationCallbacks *allocation_call_back = TVulkanAllocator::Instance()->GetVkAllocationCallbacks();
    std::cout << "0.0.2" << std::endl;

    VkResult result = Turbo::Core::vkCreateInstance(&vkinstance_create_info, allocation_call_back, &instance);
    std::cout << "0.0.3" << std::endl;

    if (result == VkResult::VK_SUCCESS)
    {
        std::cout << "0.0.4" << std::endl;

        PFN_vkDestroyInstance pfn_vk_destroy_instance = TVulkanLoader::Instance()->LoadInstanceFunction<PFN_vkDestroyInstance>(instance, "vkDestroyInstance");//it will get vkDestroyInstance function

        std::cout << "0.0.5" << std::endl;

        if (pfn_vk_destroy_instance == nullptr)
        {
            std::cout << "pfn_vk_destroy_instance::nullptr" << std::endl;
        }
        else
        {
            std::cout << "pfn_vk_destroy_instance::not nullptr" << std::endl;
        }
        pfn_vk_destroy_instance(instance, allocation_call_back);
        std::cout << "0.0.6" << std::endl;

        return true;
    }
    std::cout << "0.0.7" << std::endl;

    return false;
}

```
If I independent execution the Vulkan app, the output will be:
```
0.0.0
0.0.1
0.0.2
0.0.3
0.0.4
0.0.5
pfn_vk_destroy_instance::not nullptr
0.0.6
```
If I run it in `RenderDoc V1.25` by click `Launch` button , the output will be:
```
0.0.0
0.0.1
0.0.2
0.0.3
0.0.4
0.0.5
pfn_vk_destroy_instance::not nullptr
```
then the `Vulkan` app will crach when call `pfn_vk_destroy_instance(...)`

## Environment

* RenderDoc version: 1.25
* Operating System: Windows
* Graphics API: Vulkan

* GPU: RTX1060
* Driver Version,: 528.49

**Comment by baldurk** [comment]
There must be something more here since it is obvious that almost universally every vulkan. I can trivially show that a vulkan program does not crash when destroying the instance.

Your reproduction case very significantly uses an allocation callback but does not provide any source for it. Trying with it set to NULL does not show any problems when I try to capture the program. Does this reproduce if you don't specify an allocation callback? Can you please provide a reproducible test case that is complete and can be used to reproduce the problem?

Please provide more information: Does this happen with a simple known program such as vkcube? Do you have any other software that may be injecting itself into graphical applications such a screen recording, overlay, statistics, etc? If you attach a debugger to your program can you see where it is crashing? Can you reproduce this on multiple systems or only on the one you have tested with?

**Comment by FuXiii** [comment]
The struct about `TVulkanAllocator::Instance()->GetVkAllocationCallbacks()`
```mermaid
graph LR;
GetVkAllocationCallbacks["GetVkAllocationCallbacks()"]

pUserData[".pUserData = nullptr"]
pfnAllocation[".pfnAllocation = &TVulkanAllocator::Allocate"]
pfnReallocation[".pfnReallocation = &TVulkanAllocator::Reallocate"]
pfnFree[".pfnFree = &TVulkanAllocator::Free"]
pfnInternalAllocation[" .pfnInternalAllocation = &TVulkanAllocator::InternalAllocationNotification"]
pfnInternalFree[".pfnInternalFree = &TVulkanAllocator::InternalFreeNotification"]

GetVkAllocationCallbacks-->pUserData
GetVkAllocationCallbacks-->pfnAllocation
GetVkAllocationCallbacks-->pfnReallocation
GetVkAllocationCallbacks-->pfnFree
GetVkAllocationCallbacks-->pfnInternalAllocation
GetVkAllocationCallbacks-->pfnInternalFree

TVulkanAllocator::Allocate["void *VKAPI_PTR Allocate(void *pUserData, size_t size, size_t alignment, VkSystemAllocationScope allocationScope)"]

TVulkanAllocator::Reallocate["void *VKAPI_PTR TVulkanAllocator::Reallocate(void *pUserData, void *pOriginal, size_t size, size_t alignment, VkSystemAllocationScope allocationScope)"]

TVulkanAllocator::Free["void VKAPI_PTR TVulkanAllocator::Free(void *pUserData, void *pMemory)"]

pfnAllocation-->TVulkanAllocator::Allocate
pfnReallocation-->TVulkanAllocator::Reallocate
pfnFree-->TVulkanAllocator::Free
pfnInternalAllocation-->Donoting_pfnInternalAllocation["do nothing"]
pfnInternalFree-->Donoting_pfnInternalFree["do nothing"]

TVulkanAllocator::Allocate-->TAllocator::Allocate["return _aligned_malloc(size, alignment);"]
TVulkanAllocator::Reallocate-->TAllocator::Reallocate["return _aligned_realloc(pOriginal, size, alignment);"]
TVulkanAllocator::Free-->TAllocator::Free["return _aligned_free(pMemory);"]
```
I had tested for what you suggested, it worked in `RenderDoc` if I just return `nullptr` when call `TVulkanAllocator::Instance()->GetVkAllocationCallbacks()`. I am very thanks for you, I had show what the `VkAllocationCallbacks` struct will be in my project, please check it, thanks! thanks! thanks!

(●'◡'●)

**Comment by baldurk** [comment]
I'm not sure I quite understand what you're saying. You mean that if you don't pass the allocation callbacks then it works? If so that suggests it might be a bug in your callbacks, I'd suggest attaching a debugger and seeing where the crash happens exactly.

It might mean there's a bug in RenderDoc with handling of allocation callbacks, since they are basically never used as they are more-or-less pointless. For me to test that you need to actually give me a sample I can run as I asked above, either as source or binary - I do not understand what that graph means but it's not code that I can compile.

**Comment by FuXiii** [comment]
[PureHelloTriangle.zip](https://github.com/baldurk/renderdoc/files/10830957/PureHelloTriangle.zip) <-the `PureHelloTriangle.exe`
Ok, the project [Turbo](https://github.com/FuXiii/Turbo) is open source
```
git clone https://github.com/FuXiii/Turbo
cd ./Turbo
git submodule init
git submodule update
mkdir build
cd build
cmake -DVMA_STATIC_VULKAN_FUNCTIONS=0 -DVMA_DYNAMIC_VULKAN_FUNCTIONS=1 -DKTX_FEATURE_LOADTEST_APPS=OFF -DKTX_FEATURE_DOC=OFF -DKTX_FEATURE_STATIC_LIBRARY=ON ..
```
then into `IDE` or `cmake build` to build the project it will output `PureHelloTriangle` application. and the `PureHelloTriangle` source code in `./Turbo/samples/PureHelloTriangle.cpp`
It will crashed at https://github.com/FuXiii/Turbo/blob/4949d30e58358eddbf81ffd8666de2cba7b9d1ef/engine/core/src/TInstance.cpp#L334

by my test when call `pfn_vk_destroy_instance` it will call `_aligned_free`, and the crash point is at `_aligned_free(...)` in https://github.com/FuXiii/Turbo/blob/4949d30e58358eddbf81ffd8666de2cba7b9d1ef/engine/core/src/TAllocator.cpp#L77

and the `_aligned_free(...)` is at:
![image](https://user-images.githubusercontent.com/40132375/221354999-c517055a-134c-4df8-8966-b91ef2fae986.png)

（○｀ 3′○）

**Comment by baldurk** [comment]
I can't reproduce this with the executable provided. I tested on three different GPUs, from NV/AMD/Intel but none of them crashed when run from RenderDoc v1.25. I'm not sure exactly when it's mean to crash, but I was able to launch the program and then close it with no problems.

Have you tried updating to the latest drivers to see if that fixes your issue? are you able to reproduce this on any other system than your own? Is it possible that your executable behaves differently on other PCs? If you're able to produce a simple and deterministic sample I can test again, or if I can provide any information about the running to ensure it has actually hit the same case that you are seeing.
